pkg/crow/crow.go

```diff
@@ -0,0 +1,309 @@
+
+	f.DurationVar(&c.QueryDuration, "query-timeout", DefaultConfig.QueryTimeout, "timeout for querying")
+	f.DurationVar(&c.QueryDuration, "query-duration", DefaultConfig.QueryDuration, "time before and after sample to search")
+	f.DurationVar(&c.QueryDuration, "query-step", DefaultConfig.QueryStep, "step between samples when searching")
+
+	f.IntVar(&c.MaxValidations, "max-validations", DefaultConfig.MaxValidations, "Maximum number of times to try validating a sample")
+	f.DurationVar(&c.MaxTimestampDelta, "max-timestamp-delta", DefaultConfig.MaxTimestampDelta, "maximum difference from the stored timestamp from the validating sample to allow")
+	f.Float64Var(&c.ValueEpsilon, "sample-epsilon", DefaultConfig.ValueEpsilon, "maximum difference from the stored value from the validating sample to allow")
+}
+
+// DefaultConfig holds defaults for Crow settings.
+var DefaultConfig = Config{
+	MaxValidations:  5,
+	GenerateSamples: 10,
+
+	QueryTimeout:  150 * time.Millisecond,
+	QueryDuration: 2 * time.Second,
+	QueryStep:     100 * time.Millisecond,
+
+	// MaxTimestampDelta is set to 750ms to allow some buffer for a slow network
+	// before the scrape goes through.
+	MaxTimestampDelta: 750 * time.Millisecond,
+	ValueEpsilon:      0.0001,
+}
+
+// Crow is a collectness checker that validates scraped metrics reach a
+// Prometheus-compatible server with the same values and roughly the same
+// timestamp.
+//
+// Crow exposes two sets of metrics:
+//
+// 1. Test metrics, where each scrape generates a validation job.
+// 2. State metrics, exposing state of the Crow checker itself.
+//
+// These two metrics should be exposed via different endpoints, and only state
+// metrics are safe to be manually collecetd from.
+//
+// Collecting from the set of test metrics generates a validation job, where
+// Crow will query the Prometheus API to ensure the metrics that were scraped
+// were written with (approximately) the same timestamp as the scrape time
+// and with (approximately) the same floatnig point values exposed in the
+// scrape.
+//
+// If a set of test metrics were not found and retries have been exhausted,
+// or if the metrics were found but the values did not match, the error
+// counter will increase.
+type Crow struct {
+	cfg Config
+	m   *metrics
+
+	promClient promapi.API
+
+	wg   sync.WaitGroup
+	quit chan struct{}
+
+	pendingMtx sync.Mutex
+	pending    []*sample
+	sampleCh   chan []*sample
+}
+
+// New creates a new Crow.
+func New(cfg Config) (*Crow, error) {
+	c, err := newCrow(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	c.wg.Add(1)
+	go c.runLoop()
+	return c, nil
+}
+
+func newCrow(cfg Config) (*Crow, error) {
+	cli, err := api.NewClient(api.Config{
+		Address: cfg.PrometheusAddr,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to create prometheus client: %w", err)
+	}
+
+	c := &Crow{
+		cfg:        cfg,
+		m:          newMetrics(),
```

Comment:
I may have missed it, but are these metrics registered at any point?

Reply:
Nope, I'm trying a different pattern here. Instead of registering metrics directly, Crow exposes its two collectors (internal state & test sample generator) via a method. This means the caller has to register Crow's metrics, but it allows the caller to arbitrarily register and unregister the metrics at runtime.